### ISSUE TITLE
.mts file changing transform behavior despite ts loader

### ISSUE THREAD

**Issue by privatenumber** [evanw/esbuild#1932]
## Description

When using the Transform API with the following configuration, passing in a `filePath` that ends in `.mts` vs `.ts` seems to yield different code:
```
{
	target: 'node12',
	format: 'cjs',
	loader: 'ts',
	sourcefile: filePath,
}
```

Specifically, when the file ends in `.mts`, `1` is passed into the `isNodeMode` parameter in `__toESM` .

This is problematic when independently transforming two files where one imports the other. The imported module appears to get wrapped in another `{ default: ... }` object.

### Source files
_./index.mts_
```ts
import esm from './esm.mjs';

console.log(esm);
```

_./esm.mts_
```ts
export default 1234;
```

### Output (Unexpected behavior)

<details>
<summary><i>./index.mts</i></summary>

```js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var import_esm = __toESM(require("./esm.mjs"), 1);
console.log(import_esm.default);
```

</details>

<details>
<summary><i>./esm.mts</i></summary>

```js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var esm_exports = {};
__export(esm_exports, {
  default: () => esm_default
});
var esm_default = 1234;
module.exports = __toCommonJS(esm_exports);
```
</details>

Execution output

```
{ default: [Getter] }
```

### Output without `.mts` extension (Expected behavior)
Passing in `filePath.replace('.mts', '.ts')` to `sourcefile`.

<details>
<summary><i>./index.mts</i></summary>

```js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toESM = (module2, isNodeMode) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", !isNodeMode && module2 && module2.__esModule ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var import_esm = __toESM(require("./esm.mjs"));
console.log(import_esm.default);
```

</details>

<details>
<summary><i>./esm.mts</i></summary>

```js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var esm_exports = {};
__export(esm_exports, {
  default: () => esm_default
});
var esm_default = 1234;
module.exports = __toCommonJS(esm_exports);
```
</details>

Execution output

```
1234
```

**Comment by hyrious** [comment]
Maybe it should turn off `isNodeMode` if the import path ends with `.mjs`. But what about other extensions? What about the `esm.mjs` is actually an npm package? (e.g. the name becomes `esm`, can not know its format)

This behavior is because when you're targeting cjs format, external files are treated as cjs too, unless you can specify what format they are (#1915). The node behavior will use one <q>default</q> export to wrap the whole cjs module, despite of whether that module having defined `__esModule`. This matches what node does:

```js
// esm.js (simplified)
module.exports = { __esModule: true, default: 1234 }

// test.mjs
import esm from "./esm.js"
console.log(esm) // => { __esModule: true, default: 1234 }
```

**Comment by evanw** [comment]
> Do you think tampering with the filename to replace the `.mts` extension with `.ts` is the best way forward with this?

TL;DR: Yes.

Babel's implementation of importing CommonJS into ESM works fine with code that has been converted from ESM into CommonJS. However, node's implementation of the same thing deliberately breaks this case by ignoring the `default` export and setting the `default` export to the whole export object instead. This is an unfortunate mess. Some code is written with node's behavior in mind while other code is written with ESM behavior in mind, and esbuild needs some signal to decide which behavior to use.

Both Webpack and esbuild use the `.mjs` file extension as a signal that the code is expecting node's behavior instead of ESM behavior. In addition, esbuild uses the `.mts` extension as a signal that you are going to be compiling the file to `.mjs`, in which case you are also expecting node's behavior. This is what the `isNodeMode` parameter means.

So you should use the `.ts` extension instead of the `.mts` extension to tell esbuild that you don't want the code to behave like how `.mjs` files behave in node.

**Comment by privatenumber** [comment]
Thank you for the guidance and careful explanation, as always!

Closing since expected behavior.
